dask/array/core.py

```diff
@@ -1219,19 +1219,15 @@ def __array__(self, dtype=None, **kwargs):
         return x
 
     def __array_function__(self, func, types, args, kwargs):
-        import dask.array as module
+        if func not in _HANDLED_FUNCTIONS:
+            warnings.warn(
+                "`{}` is not implemented by dask, explicitly "
+                "coerce your array (e.g. with numpy.asarray) "
+                "to silence this warning".format(func.__name__)
+            )
+            return func._implementation(*args, **kwargs)
```

Comment:
This seems necessary to avoid infinite recursion. @shoyer are you comfortable with this? If not, any better suggestions?

Reply:
This is definitely not a great idea -- numpy reserves the right to break this implementation detail.

I would suggest using introspection instead. I'm sure there are a few edge cases this would miss, but I think this solution would work 99% of the time 
```python
import dask.array as da
import numpy as np

def maybe_compute(obj):
  if isinstance(obj, da.Array):
    return np.array(obj)
  elif isinstance(obj, tuple):
    return tuple(maybe_compute(x) for x in obj)
  elif isinstance(obj, list):
    return list(maybe_compute(x) for x in obj)
  else:
    return obj

def compute_dask_args(args, kwargs):
  args = maybe_compute(args)
  kwargs = {k: maybe_compute(v) for k, v in kwargs.items()}
  return args, kwargs

# inside __array_function__
args, kwargs = compute_dask_args(args, kwargs)
return func(args, kwargs)
```

The only case this would fail is if you use weird sequence types, e.g,. if you pass some custom sequence (not a list or tuple) full of dask arrays into `np.concatenate`.

Reply:
> I'm sure there are a few edge cases this would miss, 

Yeah that was my hesitation. Well, I guess this is the kind of thing that made you want to expose `_implementation`, but we decided not to.....

Okay, will change to your suggestion.

Reply:
done